src/EventSubscriber/OrderEventSubscriber.php

```diff
@@ -101,40 +116,63 @@ protected function giftcardPurchase(OrderInterface $order) {
-          ]);
-          $giftcard->save();
-
-          $transaction = $this->entityTypeManager->getStorage('commerce_giftcard_transaction')->create([
-            'giftcard' => $giftcard->id(),
-            'amount' => $amount,
-            'reference_type' => $item->getEntityTypeId(),
-            'reference_id' => $item->id(),
-            'comment' => 'Bought @product.',
-            'variables' => ['@product' => $purchased_entity->label()]
-          ]);
-          $transaction->save();
-        }
+      if (!$purchased_entity) {
+        return;
+      }
+      $giftcard_amount = $this->getAmountFromItem($item);
+      if (!$giftcard_amount instanceof Price) {
+        return;
+      }
+      $codes = $this->codeGenerator->generateCodes($purchased_entity->get('commerce_giftcard_type')->entity, $item->getQuantity());
+
+      for ($i = 0; $i < $item->getQuantity(); $i++) {
+        // Create a giftcard and then add the balance as a transaction to
+        // store the reference to this order.
+        $giftcard = $this->entityTypeManager->getStorage('commerce_giftcard')->create([
+          'type' => $purchased_entity->get('commerce_giftcard_type')->target_id,
+          'code' => $codes[$i],
+          'balance' => new Price(0, $giftcard_amount->getCurrencyCode()),
+          'uid' => $order->getCustomerId(),
+          // Set the stores to the stores that the purchasable entity can be
+          // bought not just the one from the order.
+          // @todo Make this configurable, add an event for the giftcard and
+          //   transaction being created?
+          'stores' => $purchased_entity->getStores(),
+        ]);
+        $giftcard->save();
+
+        $transaction = $this->entityTypeManager->getStorage('commerce_giftcard_transaction')->create([
+          'giftcard' => $giftcard->id(),
+          'amount' => $giftcard_amount,
+          'reference_type' => $item->getEntityTypeId(),
+          'reference_id' => $item->id(),
+          'comment' => 'Bought @product.',
+          'variables' => ['@product' => $purchased_entity->label()],
+        ]);
+        $transaction->save();
       }
     }
   }
 
+  /**
+   * Helper to get a giftcard amount from an order item.
+   *
+   * @param \Drupal\commerce_order\Entity\OrderItemInterface $item
+   *   Order item.
+   *
+   * @return \Drupal\commerce_price\Price|null
+   *   An amount in the form of a price, if applicable. Or null otherwise.
+   */
+  protected function getAmountFromItem(OrderItemInterface $item) {
+    $purchased_entity = $item->getPurchasedEntity();
+    $giftcard_amount = $item->getAdjustedUnitPrice();
+    if ($purchased_entity->hasField('commerce_giftcard_amount') && $purchased_entity->hasField('commerce_giftcard_type') && !$purchased_entity->get('commerce_giftcard_amount')->isEmpty()) {
+      $giftcard_amount = $purchased_entity->get('commerce_giftcard_amount')->first()->toPrice();
+    }
+
+    $event = new GiftcardAmountCalculateEvent($giftcard_amount);
```

Comment:
shouldn't we at least pass the order item as context to the event as well, otherwise that's pretty pointless? 

Also, now you even made the existence of those fields optional, so this will return the amount of _any_ order item?

We definitely need at least a non-empty commerce_giftcard_type as detection to return anything as that's used later on. And only if that is the case we can start with the giftcard_amount default and use the commerce_giftcard_amount field if set.

Lets also add a description to the price field in \Drupal\commerce_giftcard\Plugin\Commerce\EntityTrait\GiftcardPurchase::buildFieldDefinitions() so that it says that if empty, the price is used?

Not 100% sure which price is the right one to use, the adjusted or non-adjusted one.

Reply:
>shouldn't we at least pass the order item as context to the event as well, otherwise that's pretty pointless?
Heh, very good point

>Also, now you even made the existence of those fields optional, so this will return the amount of any order item?
Also good point. Adding that check.

>Not 100% sure which price is the right one to use, the adjusted or non-adjusted one.
Also not 100% sure, but I feel 1:1 with order item (as paid) makes the most sense. Otherwise it would be free money (for example if the whole store had a 50% sale, and you purchased a 100$ giftcard for 50$). User error, for sure, but still...